test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/CharArrayWriterTest.scala

```diff
@@ -0,0 +1,181 @@
+import org.junit.Assert._
+import org.junit.Test
+import org.scalajs.testsuite.utils.AssertThrows._
+
+class CharArrayWriterTest {
+  private[io] val hw: Array[Char] = Array('H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd')
+
+  private def withClose[T <: AutoCloseable](closeable: T)(fn: T => Unit): Unit = {
+    fn(closeable)
+
+    if (closeable != null) closeable.close()
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#CharArrayWriter(int)
+    */
+  @Test def should_construct_with_correct_size_when_supplied(): Unit = withClose(new CharArrayWriter(90)) { cw =>
+    assertEquals("Created incorrect writer", 0, cw.size)
+  }
+
+
+  /**
+    * @tests java.io.CharArrayWriter#CharArrayWriter(int)
+    */
+  @Test def should_construct_with_exception(): Unit = {
+    assertThrows(classOf[IllegalArgumentException], new CharArrayWriter(-1))
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#CharArrayWriter()
+    */
+  @Test def should_construct_with_correct_size_default(): Unit = withClose(new CharArrayWriter) { cw =>
+    assertEquals("Created incorrect writer", 0, cw.size)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#close()
+    */
+  @Test def should_support_close(): Unit = withClose(new CharArrayWriter) { cw =>
+    cw.close()
+    cw.close() // no-op
+    assertEquals("Size should work after close", 0, cw.size)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#flush()
+    */
+  @Test def should_support_flush(): Unit = withClose(new CharArrayWriter) { cw =>
+    cw.flush()
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#reset()
+    */
+  @Test def should_support_reset(): Unit = withClose(new CharArrayWriter) { cw =>
+    cw.write("HelloWorld", 5, 5)
+    cw.reset()
+    cw.write("HelloWorld", 0, 5)
+
+    assertArrayEquals("Reset failed to reset buffer", "Hello".toCharArray, cw.toCharArray)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#size()
+    */
+  @Test def should_support_size(): Unit = withClose(new CharArrayWriter) { cw =>
+    assertEquals("Returned incorrect size", 0, cw.size)
+    cw.write(hw, 5, 5)
+    assertEquals("Returned incorrect size", 5, cw.size)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#toCharArray()
+    */
+  @Test def should_support_toCharArray(): Unit = withClose(new CharArrayWriter) { cw =>
+    cw.write("HelloWorld", 0, 10)
+    assertArrayEquals("toCharArray failed to return correct array", "HelloWorld".toCharArray, cw.toCharArray)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#toString()
+    */
+  @Test def should_support_toString(): Unit = withClose(new CharArrayWriter) { cw =>
+    cw.write("HelloWorld", 5, 5)
+    assertEquals("Returned incorrect string", "World", cw.toString)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#write(char[], int, int)
+    */
+  @Test def should_support_write_char_array(): Unit = withClose(new CharArrayWriter) { cw =>
+    cw.write(hw, 5, 5)
+    assertEquals("Writer failed to write correct chars", "World", new String(cw.toCharArray, 0, 5))
```

Comment:
The `String` constructor truncates its input, which weakens the test (`toCharArray` could return a longer array than expected without making the test fail). It's better to use `assertArrayEquals` and convert the expected value to an array instead:
```suggestion
    assertArrayEquals("Writer failed to write correct chars", "World".toCharArray, cw.toCharArray)
```